Fix 4314 - Items in settings errors transient not displayed on settings update

iCaspar opened the pull request:
## Description

Changes our settings error process to use `add_settings_error` instead of the settings_errors transient.

Fixes #4314

## Type of change
- [X] Bug fix (non-breaking change which fixes an issue)

## Is the solution different from the one proposed during the grooming?

Same as grooming.

## How Has This Been Tested?

Checked that errors are being displayed on local development environment.

# Checklist:

Please delete the options that are not relevant.

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

iCaspar committed:
Use add_settings_error() to set error messages

iCaspar committed:
Remove some invisible characters

iCaspar committed:
Fix foreach spacing

Mai-Saad commented:
@iCaspar please find my notes/Qs below while exploring the PR:

1- `PHP Warning:  preg_match(): Compilation failed: missing closing parenthesis` will exist in logs for  Defer, and Delay JS => just confirming that the current PR won't fix this problem, right?

2- Notice is not displayed for the following fields although `gtag(` was removed: Never Cache Url(s), Always Purge URL(s), Preload Cache, and Preload Fonts => will we handle their notice in the current PR?

3- Although  `gtag(` is removed and notice is displayed for combine js, we have this warning in debug.log `PHP Warning:  preg_match(): Compilation failed: missing closing parenthesis at offset 928 in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Optimization/Minify/JS/Combine.php on line 914` (same warning on the trunk while notice not displayed) => Will this be fixed with 1 related issue?

4- Q: which fields shall escape invalid regex by adding `\` and which fields shall remove invalid regex totally?

iCaspar commented:
On 1 and 3, indeed, this will not fix any warnings. This PR will ONLY address that the message is not being displayed when an invalid entry was removed.

On 2 and 4, the notice should be displayed for an invalid regex that is removed in the following fields:
```
		'exclude_css'       => __( 'Excluded CSS Files', 'rocket' ),
		'exclude_inline_js' => __( 'Excluded Inline JavaScript', 'rocket' ),
		'exclude_js'        => __( 'Excluded JavaScript Files', 'rocket' ),
		'cache_reject_uri'  => __( 'Never Cache URL(s)', 'rocket' ),
		'cache_reject_ua'   => __( 'Never Cache User Agent(s)', 'rocket' ),
		'cache_purge_pages' => __( 'Always Purge URL(s)', 'rocket' ),
		'cdn_reject_files'  => __( 'Exclude files from CDN', 'rocket' ),
```

Mai-Saad commented:
@iCaspar Notice is displayed for all mentioned fields except `'cache_reject_uri'  => __( 'Never Cache URL(s)', 'rocket' ),` and `'cache_purge_pages' => __( 'Always Purge URL(s)', 'rocket' ),` Can we handle those?

- Other Note: If we have multiple invalid regex, the notice is displayed only on the last item.

Example: When excluding the following from CSS :
```
h(.*)apply(.*)(
h(.*)apply(.*))
gtag(
(test
```
although all items were removed from the field, the notice is :
![lastErrorOnly](https://user-images.githubusercontent.com/76941962/142579798-9df3113d-5407-461a-ae9f-eed36e1937b3.jpg) 

Shouldn't we display all removed items in the notice?

iCaspar commented:
@Mai-Saad @piotrbak 

> Shouldn't we display all removed items in the notice?

Yes!
I've added a bit to make that happen now. Should be like this:
<img width="467" alt="Screen Shot 2021-11-22 at 10 59 52 AM" src="https://user-images.githubusercontent.com/7462998/142894331-77c2a132-d040-4e9f-898d-1d6ba0ce7821.png">

Also, looks like in the case of 
* Never Cache URLs
* Always Purge URLs, and
* Exclude files from CDN
Turns out we are also, _**in a different sanitization**_, removing anything from those fields that is not an actual file -- regardless of any regex.
It will probably be better to handle these 3 in a separate issue/pr.

Also, I've added the DelayJS to the pattern fields that will display here based on regex removal. So at least the "Warning" part of #4136 should be applied here (not the other things about notices in logs, etc.)

iCaspar committed:
Add DelayJS and display multiple errors per field

iCaspar committed:
Fix spacing

Mai-Saad reviewed (commented):
As per discussions and latest analysis, this PR works as expected to display notice for all mentioned fields (except Never Cache URL and Always purge URL => will be in another ticket)

Notes:
- Some notices have a typo (will be in another ticket)
- Some fields cause warning (will be in another ticket)
-  Exclusion for `test\(` is removed from js/CSS files (will be in another ticket)
- Analysis Sheet here https://docs.google.com/spreadsheets/d/1NQz8vhTpuTa9Jqaf8TcyfLmhinc0Lx59ulo_ZUR9r0c/edit?usp=sharing
